### PR TITLE
feat(functions): add map_contains_key scalar function

### PR DESCRIPTION
**Pull request by hanxuanliang**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

- Addresses part of https://github.com/datafuselabs/databend/issues/15295

## Tests

- [x] Unit Test
- [ ] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/15465)

### CODE REVIEW

**Comment by hanxuanliang** [comment]
During the execution of unit tests, I encountered a problem: 

```
```

```txt
successes:

successes:

failures:

---- scalars::map::test_map stdout ----
thread 'scalars::map::test_map' panicked at /Users/hanfox/.cargo/registry/src/rsproxy.cn-0dccff568467c15b/goldenfile-1.6.0/src/differs.rs:15:5:
assertion failed: `(left == right)`'
  left: `"ast            : map([], [])\nraw expr       : map(array(), array())\nchecked expr   : map<Array(Nothing), Array(Nothing)>(array<>(), array<>())\noptimized expr : {} :: Map(Nothing)\noutput type    : Map(..."` (truncated)
 right: `"ast            : map([], [])\nraw expr       : map(array(), array())\nchecked expr   : map<Array(Nothing), Array(Nothing)>(array<>(), array<>())\noptimized expr : {} :: Map(Nothing)\noutput type    : Map(..."` (truncated)

Differences (-left|+right):
 output type    : Boolean
 output domain  : {TRUE}
 output         : true

+
```

my question is: How should I read this error? And I don't know if my unit test writing is correct.

**Comment by hanxuanliang** [comment]
@b41sh 

When performing unit tests, I encountered a problem:

```txt
cargo test --package databend-common-functions --test it -- scalars::map::test_map --exact --show-output
```

error:
```txt
running 1 test
test scalars::map::test_map ... FAILED

successes:

successes:

failures:

---- scalars::map::test_map stdout ----
thread 'scalars::map::test_map' panicked at /Users/hanfox/.cargo/registry/src/rsproxy.cn-0dccff568467c15b/goldenfile-1.6.0/src/differs.rs:15:5:
assertion failed: `(left == right)`'
  left: `"ast            : map([], [])\nraw expr       : map(array(), array())\nchecked expr   : map<Array(Nothing), Array(Nothing)>(array<>(), array<>())\noptimized expr : {} :: Map(Nothing)\noutput type    : Map(..."` (truncated)
 right: `"ast            : map([], [])\nraw expr       : map(array(), array())\nchecked expr   : map<Array(Nothing), Array(Nothing)>(array<>(), array<>())\noptimized expr : {} :: Map(Nothing)\noutput type    : Map(..."` (truncated)

Differences (-left|+right):
 output type    : Boolean
 output domain  : {TRUE}
 output         : true

+
```

I have two areas of doubt:

1. How should I interpret this error?
2. How can I determine if my unit tests are written correctly?

Thank you for taking the time to assist me with this task.

**Comment by b41sh** [comment]
> I have two areas of doubt:
> 
> 1. How should I interpret this error?
> 2. How can I determine if my unit tests are written correctly?
> 
> Thank you for taking the time to assist me with this task.

Hi, @hanxuanliang thanks for your contribution. The unit test is failed because you missed some blank lines, just add those lines will pass the test.
```
Differences (-left|+right):
 output type    : Map(String, Map(String, String))
 output domain  : {[{"k1"..="k2"}], [{[{"nk1"..="nk3"}], [{"new_nv1"..="nv3"}]}]}
 output         : {'k1':{'nk1':'new_nv1'}, 'k2':{'nk3':'nv3'}}

+
 ast            : map_contains_key({'k1': 'v1', 'k2': 'v2'}, 'k1')
 raw expr       : map_contains_key(map(array('k1', 'k2'), array('v1', 'v2')), 'k1')
 checked expr   : map_contains_key<T0=String, T1=String><Map(T0, T1), String>(map<T0=String, T1=String><Array(T0), Array(T1)>(array<T0=String><T0, T0>("k1", "k2"), array<T0=String><T0, T0>("v1", "v2")), "k1")
 optimized expr : true
 output type    : Boolean
 output domain  : {TRUE}
 output         : true
+
+
```
you should also modify the function list file `src/query/functions/tests/it/scalars/testdata/function_list.txt`, add the new function in the function list.
```
Differences (-left|+right):
 0 map_cat(Map(Nothing), Map(Nothing)) :: Map(Nothing)
 1 map_cat(Map(Nothing) NULL, Map(Nothing) NULL) :: Map(Nothing) NULL
 2 map_cat(Map(T0, T1), Map(T0, T1)) :: Map(T0, T1)
 3 map_cat(Map(T0, T1) NULL, Map(T0, T1) NULL) :: Map(T0, T1) NULL
+0 map_contains_key(Map(Nothing), String) :: Boolean
+1 map_contains_key(Map(T0, T1), String) :: Boolean
+2 map_contains_key(Map(T0, T1) NULL, String NULL) :: Boolean NULL
 0 map_keys(Map(Nothing)) :: Array(Nothing)
 1 map_keys(Map(T0, T1)) :: Array(T0)
 2 map_keys(Map(T0, T1) NULL) :: Array(T0) NULL
 0 map_size(Map(Nothing)) :: UInt8
```

File: src/query/functions/src/scalars/map.rs [hunk 1]
```diff
@@ -227,4 +231,22 @@ pub fn register(registry: &mut FunctionRegistry) {
         |_, _| FunctionDomain::Full,
         |map, _| map.len() as u64,
     );
+
+    registry.register_2_arg_core::<EmptyMapType, StringType, BooleanType, _, _>(
```
Comment (b41sh) [inline src/query/functions/src/scalars/map.rs:235]: The type of the second argument should be generic type. The key of the map also supports other types, such as int.

File: src/query/functions/src/scalars/map.rs [hunk 2]
```diff
@@ -227,4 +231,22 @@ pub fn register(registry: &mut FunctionRegistry) {
...
+            FunctionDomain::Domain(BooleanDomain {
+                has_false: true,
+                has_true: false,
+            })
+        },
+        |_, _, _| Value::Scalar(false),
+    );
+
+    registry
+        .register_2_arg::<MapType<GenericType<0>, GenericType<1>>, StringType, BooleanType, _, _>(
```
Comment (b41sh) [inline src/query/functions/src/scalars/map.rs:247]: ditto

File: src/query/functions/tests/it/scalars/map.rs [hunk 3]
```diff
@@ -182,6 +183,11 @@ fn test_map_keys(file: &mut impl Write) {
     );
 }
 
+fn test_map_contains_key(file: &mut impl Write) {
+    run_ast(file, "map_contains_key({'k1': 'v1', 'k2': 'v2'}, 'k1')", &[
```
Comment (b41sh) [inline src/query/functions/tests/it/scalars/map.rs:187]: add some test case with empty map and other key types.

Comment (b41sh) [reply src/query/functions/tests/it/scalars/map.rs:187]: By the way, please add a few sql logic tests to this file `tests/sqllogictests/suites/query/functions/02_0074_function_map.test`

File: src/query/functions/src/scalars/map.rs (hunk 1 above)
Comment (hanxuanliang) [reply src/query/functions/src/scalars/map.rs]: Indeed, I overlooked this type.

File: src/query/functions/tests/it/scalars/testdata/function_list.txt [hunk 4]
```diff
@@ -2440,6 +2440,7 @@ Functions overloads:
 1 map_cat(Map(Nothing) NULL, Map(Nothing) NULL) :: Map(Nothing) NULL
 2 map_cat(Map(T0, T1), Map(T0, T1)) :: Map(T0, T1)
 3 map_cat(Map(T0, T1) NULL, Map(T0, T1) NULL) :: Map(T0, T1) NULL
+0 map_contains_key(Map(T0, T1), T0) :: Boolean
```
Comment (b41sh) [inline src/query/functions/tests/it/scalars/testdata/function_list.txt:2443]: seems not correct
```
-0 map_contains_key(Map(T0, T1), T0) :: Boolean
+0 map_contains_key(Map(Nothing), T0) :: Boolean
+1 map_contains_key(Map(T0, T1), T0) :: Boolean
+2 map_contains_key(Map(T0, T1) NULL, T0 NULL) :: Boolean NULL
```

File: src/query/functions/src/scalars/map.rs [hunk 5]
```diff
@@ -227,4 +228,20 @@ pub fn register(registry: &mut FunctionRegistry) {
...
+        |_, _, _| Value::Scalar(false),
+    );
+
+    registry
+        .register_2_arg::<MapType<GenericType<0>, GenericType<1>>, GenericType<0>, BooleanType, _, _>(
+            "map_contains_key",
+            |_, _, _| FunctionDomain::Full,
+            |map, key, _| {
+                map.iter()
+                    .any(|(k, _)| format!("{:?}", k) == format!("{:?}", key))
```
Comment (b41sh) [inline src/query/functions/src/scalars/map.rs:244]: better to use `k == key`

File: src/query/functions/tests/it/scalars/testdata/function_list.txt (hunk 4 above)
Comment (hanxuanliang) [reply src/query/functions/tests/it/scalars/testdata/function_list.txt:2443]: May I ask, is it possible to deduce from the results of unit tests? Similar to when I was writing `map.txt`, I did so by examining the results of the unit tests to write the correct map.txt.

Comment (b41sh) [reply src/query/functions/tests/it/scalars/testdata/function_list.txt:2443]: You can write `map.txt` by extrapolating the results from the inputs of the unit tests, but since there are more result fields, we usually just copy the output from the run. Of course, you need to check whether the output is correct.

File: tests/sqllogictests/suites/query/functions/02_0074_function_map.test [hunk 6]
```diff
@@ -137,3 +137,51 @@ SELECT MAP_CAT(
 
 statement ok
 DROP DATABASE map_func_test
+
+# Test map_contains_key function
+statement ok
+CREATE DATABASE IF NOT EXISTS map_contains_key_test
```
Comment (b41sh) [inline tests/sqllogictests/suites/query/functions/02_0074_function_map.test:143]: use db `map_func_test`, we don't need add a new database for this function test.

File: tests/sqllogictests/suites/query/functions/02_0074_function_map.test [hunk 7]
```diff
@@ -137,3 +137,51 @@ SELECT MAP_CAT(
...
+
+statement ok
+USE map_contains_key_test
+
+query
```
Comment (b41sh) [inline tests/sqllogictests/suites/query/functions/02_0074_function_map.test:148]: this should be `statement ok`

File: tests/sqllogictests/suites/query/functions/02_0074_function_map.test [hunk 8]
```diff
@@ -137,3 +137,51 @@ SELECT MAP_CAT(
...
+CREATE TABLE t(col1 Map(String, String Null) Not Null, col2 Map(String, Int Null) Null)
+----
+CREATE TABLE t(col1 Map(String, String Null) Not Null, col2 Map(String, Int Null) Null)
```
Comment (b41sh) [inline tests/sqllogictests/suites/query/functions/02_0074_function_map.test:151]: remove this two lines, ddl statement don't need to check result.

File: tests/sqllogictests/suites/query/functions/02_0074_function_map.test [hunk 9]
```diff
@@ -137,3 +137,51 @@ SELECT MAP_CAT(
...
+
+query
+INSERT INTO t VALUES({'k1':'v1','k2':'v2','k3':null},{'a':10,'b':20}), ({'k5':'v5','k6':'v6'}, {'d':40,'e':null,'f':50}), ({}, null)
+----
+INSERT INTO t VALUES({'k1':'v1','k2':'v2','k3':null},{'a':10,'b':20}), ({'k5':'v5','k6':'v6'}, {'d':40,'e':null,'f':50}), ({}, null)
```
Comment (b41sh) [inline tests/sqllogictests/suites/query/functions/02_0074_function_map.test:156]: ditto

**Comment by b41sh** [comment]
this PR is almost OK, please fix sql logic test reviews.

**Comment by hanxuanliang** [comment]
But when I was running the SQL logic test locally, I encountered a problem:

first I need to start databend server: `make run-debug`. then an issue occurred while executing this command:

```txt
Databend Metasrv started
panicked at src/meta/client/src/lib.rs:47:28:
run `git fetch --tags` to solve this error,
    to learn more about this error, please visit https://crates.io/crates/semver: Error("unexpected character 'D' while parsing major version number")
   0: std::backtrace_rs::backtrace::libunwind::trace
             at /rustc/8ace7ea1f7cbba7b4f031e66c54ca237a0d65de6/library/std/src/../../backtrace/src/backtrace/libunwind.rs:104:5
   1: std::backtrace_rs::backtrace::trace_unsynchronized
             at /rustc/8ace7ea1f7cbba7b4f031e66c54ca237a0d65de6/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
             ...
Databend Query start failure, cause: MetaServiceError. Code: 2001, Text = meta-client dedicated runtime error: tokio::sync::oneshot::error::RecvError: channel closed while: when recv resp from MetaGrpcClient worker.
```

An error occurred when starting databend-query. I initially thought it was an issue with port 8000, but even after switching to port 8001, the same problem persisted.

**Comment by b41sh** [comment]
> But when I was running the SQL logic test locally, I encountered a problem:
> 
> first I need to start databend server: `make run-debug`. then an issue occurred while executing this command:
> 
> ```
> Databend Metasrv started
> panicked at src/meta/client/src/lib.rs:47:28:
> run `git fetch --tags` to solve this error,
>     to learn more about this error, please visit https://crates.io/crates/semver: Error("unexpected character 'D' while parsing major version number")
>    0: std::backtrace_rs::backtrace::libunwind::trace
>              at /rustc/8ace7ea1f7cbba7b4f031e66c54ca237a0d65de6/library/std/src/../../backtrace/src/backtrace/libunwind.rs:104:5
>    1: std::backtrace_rs::backtrace::trace_unsynchronized
>              at /rustc/8ace7ea1f7cbba7b4f031e66c54ca237a0d65de6/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
>              ...
> Databend Query start failure, cause: MetaServiceError. Code: 2001, Text = meta-client dedicated runtime error: tokio::sync::oneshot::error::RecvError: channel closed while: when recv resp from MetaGrpcClient worker.
> ```
> 
> An error occurred when starting databend-query. I initially thought it was an issue with port 8000, but even after switching to port 8001, the same problem persisted.

databend use github tags to calculate the version, otherwise the meta and query version is incorrect, which will cause  the startup failed.
please run the following command first and then rebuild the target
```
git fetch --tags https://github.com/datafuselabs/databend.git
```

**Review by b41sh — APPROVED** [review]
LGTM @hanxuanliang Thanks for your contribution.

**Comment by hanxuanliang** [comment]
@b41sh Thank you for your help, I can successfully complete the development.. At the same time, during the development, I feel that the develop process of scalar func is not very clear in the documentation. Is there also support to contribute?
